src/engine/media/commands.ts

```diff
@@ -1,27 +1,105 @@
 import {Fetch} from "hurdak"
-import {generatePrivateKey} from "nostr-tools"
+import {generatePrivateKey, type Event} from "nostr-tools"
 import {signer} from "src/engine/session/derived"
 import {buildEvent} from "src/engine/network/utils"
+import crypto from "crypto"
+import { get } from "svelte/store"
+
+export const uploadToMediaProvider = async (url: string, body: FormData): Promise<string> => {
 
-export const uploadToNostrBuild = async body => {
-  const $signer = signer.get()
-  const url = "https://nostr.build/api/v2/upload/files"
-  const template = buildEvent(27235, {
-    tags: [
-      ["u", url],
-      ["method", "POST"],
-    ],
-  })
 
-  const event = await ($signer.canSign()
-    ? $signer.signAsUser(template)
-    : $signer.signWithKey(template, generatePrivateKey()))
+  //Convert MediaProvider URL to NIP96 upload URL
+  url = await getMediaProviderURL(url);
 
-  return Fetch.fetchJson(url, {
+  //Prepare NIP98 event
+  const event : Event = await prepareNIP98Event(url, "POST", body)
+  const response = await Fetch.fetchJson(url, {
     body,
     method: "POST",
     headers: {
       Authorization: `Nostr ${btoa(JSON.stringify(event))}`,
-    },
+  }})
+  console.debug("response", response)
+
+  // If the media provider use delayed processing, we need to wait for the processing to be done
+  if(response.processing_url.length > 0){
+
+    //Prepare NIP98 event	for delayed processing. TODO. If the delayed processing is too long, the event "created_at" maybe expire. It depends on the media provider.
+    const event_processing : Event = await prepareNIP98Event(response.processing_url, "GET", body)
+    let processing_response = await Fetch.fetchJson(response.processing_url, {
+      method: "GET",
+      headers: {
+        Authorization: `Nostr ${btoa(JSON.stringify(event_processing))}`,
+      }})
+    console.debug("processing_response", processing_response)
+
+    let timeout : number = 0
+    while(processing_response.status == "processing"){
+      if (timeout == 120){ // 2 minutes of timeout. TODO, make it configurable
+        console.debug("Processing timeout reached");
+        return "";
+      }
+      processing_response = await Fetch.fetchJson(response.processing_url, {
+        method: "GET",
+        headers: {
+          Authorization: `Nostr ${btoa(JSON.stringify(event_processing))}`,
+        }})
+
+      if (processing_response.status == "success"){
+        console.debug("Processing done");
+        return findUrlTag(processing_response.nip94_event.tags)
+      }
+      console.debug("Status", processing_response.status, "(" + processing_response.percentage + "%)")
+      await new Promise(resolve => setTimeout(resolve, 1000)) // 1 second of delay
+      timeout++
+    }
+  }
+    
+  if(response.nip94_event.tags.length > 0){
+    return findUrlTag(response.nip94_event.tags)
+  }else{
+    return ""
+  }
```

Comment:
This can be simplifed to remove the else by using `Tags` and returning null/undefined if no match

Reply:
I have changed it, but there should be an output with "" if the response does not contain a NIP94 event (server malformed for example). 

![image](https://github.com/coracle-social/coracle/assets/125748180/8d067f53-914f-4ba5-988c-ca847c1432db)

Reply:
@staab it is ok for you? 😀